spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala

```diff
@@ -55,11 +55,19 @@ class Adagrad[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeri
     val lrd = _config.get[Double]("learningRateDecay").getOrElse(0.0)
     val nevals = _state.get[Int]("evalCounter").getOrElse(0)
     val wd = config.get[Double]("weightDecay").getOrElse(0.0)
+    val l1wd = config.get[Double]("l1WeightDecay").getOrElse(0.0)
+    val l2wd = config.get[Double]("l2WeightDecay").getOrElse(wd)
+    val weightsSignBuffer = _state.get[Tensor[T]]("weightSigns").getOrElse(Tensor(parameter.size()))
+    _state.update("weightSigns", weightsSignBuffer)
 
     val (fx, dfdx) = feval(parameter)
 
-    if (wd != 0) {
-      dfdx.add(ev.fromType[Double](wd), parameter)
+    val l1Regularizer = weightsSignBuffer.copy(parameter).sign()
+    val l2Regularizer = parameter
+
+    if (l1wd != 0 || l2wd != 0) {
```

Comment:
if l1wd/l2wd is zero. You can skip the l1/l2 weigth decay. But here you always do it.